runtime/src/iree/hal/drivers/vulkan/native_allocator.cc

```diff
@@ -0,0 +1,368 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cstddef>
+#include <cstring>
+
+#include "iree/base/api.h"
+#include "iree/hal/drivers/vulkan/base_buffer.h"
+#include "iree/hal/drivers/vulkan/dynamic_symbols.h"
+#include "iree/hal/drivers/vulkan/native_buffer.h"
+#include "iree/hal/drivers/vulkan/status_util.h"
+
+using namespace iree::hal::vulkan;
+
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_ALLOCATION_TRACKING
+static const char* IREE_HAL_VULKAN_NATIVE_ALLOCATOR_ID = "Vulkan/Native";
+#endif  // IREE_TRACING_FEATURE_ALLOCATION_TRACKING
+
+typedef struct iree_hal_vulkan_native_allocator_t {
+  iree_hal_resource_t resource;
+  VkDeviceHandle* logical_device;
+  iree_hal_device_t* device;  // unretained to avoid cycles
+  iree_allocator_t host_allocator;
+
+  // Cached from the API to avoid additional queries in hot paths.
+  VkPhysicalDeviceProperties device_props;
+  VkPhysicalDeviceMemoryProperties memory_props;
+
+  // Used to quickly look up the memory type index used for a particular usage.
+  iree_hal_vulkan_memory_types_t memory_types;
+
+  IREE_STATISTICS(iree_hal_allocator_statistics_t statistics;)
+} iree_hal_vulkan_native_allocator_t;
+
+namespace {
+extern const iree_hal_allocator_vtable_t
+    iree_hal_vulkan_native_allocator_vtable;
+}  // namespace
+
+static iree_hal_vulkan_native_allocator_t*
+iree_hal_vulkan_native_allocator_cast(iree_hal_allocator_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_vulkan_native_allocator_vtable);
+  return (iree_hal_vulkan_native_allocator_t*)base_value;
+}
+
+static void iree_hal_vulkan_native_allocator_destroy(
+    iree_hal_allocator_t* IREE_RESTRICT base_allocator);
+
+extern "C" iree_status_t iree_hal_vulkan_native_allocator_create(
```

Comment:
Some day, we are gonna write this whole thing to pure C, right? :)

Reply:
yes, really really yes x_x